### PR TITLE
Adds invite users modal [SCI-1498]

### PR DESCRIPTION
**Pull request by mlorb**
[SCI-1498](https://biosistemika.atlassian.net/browse/SCI-1498)

### CODE REVIEW

File: app/javascript/src/components/InviteUsersModal/components/InviteUsersResults.jsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import PropTypes from 'prop-types';
```
Comment (ZmagoD) [inline app/javascript/src/components/InviteUsersModal/components/InviteUsersResults.jsx]: same here

File: app/javascript/src/components/InviteUsersModal/index.jsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
```
Comment (ZmagoD) [inline app/javascript/src/components/InviteUsersModal/index.jsx]: same here

File: app/javascript/src/components/InviteUsersModal/index.jsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import InviteUsersForm from './components/InviteUsersForm';
+import InviteUsersResults from './components/InviteUsersResults';
+import InviteUsersButton from './components/InviteUsersButton';
+
+const StyledButtonToolbar = styled(ButtonToolbar)`
+	float: right;
+`;
+
+class InviteUsersModal extends Component {
+	constructor() {
```
Comment (ZmagoD) [inline app/javascript/src/components/InviteUsersModal/index.jsx]: you have to pass "props2 in the constructor as param and then pass it to super

File: app/services/client_api/invitations_service.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      @role = args[:role]
+
+      raise ClientApi::CustomInvitationsError unless @emails && @team && @role
+      @emails && @emails.empty? { raise ClientApi::CustomInvitationsError }
+      @role && !UserTeam.roles.keys.include?(@role) { raise ClientApi::CustomInvitationsError }
+    end
+
+    def invitation
+      invite_results = []
+
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/services/client_api/invitations_service.rb]: Can you make this method to look nicer :smile:  maybe extract some chunks of code in separate private methods

File: app/javascript/src/components/InviteUsersModal/components/InviteUsersButton.jsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import PropTypes from 'prop-types';
```
[the remark this answers is not in the document]
Comment (mlorb) [reply app/javascript/src/components/InviteUsersModal/components/InviteUsersButton.jsx]: :+1:

File: app/javascript/src/components/InviteUsersModal/components/InviteUsersForm.jsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import PropTypes from 'prop-types';
```
[the remark this answers is not in the document]
Comment (mlorb) [reply app/javascript/src/components/InviteUsersModal/components/InviteUsersForm.jsx]: :+1:

File: app/javascript/src/components/InviteUsersModal/components/InviteUsersResults.jsx (hunk 1 above)
Comment (mlorb) [reply app/javascript/src/components/InviteUsersModal/components/InviteUsersResults.jsx]: :+1:

File: app/javascript/src/components/InviteUsersModal/index.jsx (hunk 2 above)
Comment (mlorb) [reply app/javascript/src/components/InviteUsersModal/index.jsx]: :+1:

File: app/services/client_api/invitations_service.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (mlorb) [reply app/services/client_api/invitations_service.rb]: :+1:

File: app/javascript/src/components/InviteUsersModal/components/InviteUsersForm.jsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import { FormattedMessage } from 'react-intl';
+import { FormGroup, HelpBlock } from 'react-bootstrap';
+import TagsInput from 'react-tagsinput';
+
+import { INVITE_USERS_LIMIT } from '../../../config/constants/numeric';
+
+const InviteUsersForm = props =>
+	<FormGroup controlId="form-invite-user">
+		<p>
+			<FormattedMessage id="invite_users.input_text" values={{team: props.teamName}} />
```
[the remark this answers is not in the document]
Comment (mlorb) [reply app/javascript/src/components/InviteUsersModal/components/InviteUsersForm.jsx]: I need <p> tag for styling

File: app/javascript/src/components/InviteUsersModal/index.jsx (hunk 3 above)
Comment (mlorb) [reply app/javascript/src/components/InviteUsersModal/index.jsx]: :+1:

File: app/javascript/src/components/InviteUsersModal/index.jsx [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      .put(INVITE_USERS_PATH, {
+        user_role: role,
+        emails: this.state.inputTags,
+        team_id: this.props.team.id
+      })
+      .then(({ data }) => {
+        this.setState({ inviteResults: data });
+        this.setState({ showInviteUsersResults: true });
+      })
+      .catch(error => {});
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/javascript/src/components/InviteUsersModal/index.jsx]: :banana:

File: app/javascript/src/components/InviteUsersModal/index.jsx [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+  inviteAs(role) {
+    axios
+      .put(INVITE_USERS_PATH, {
+        user_role: role,
+        emails: this.state.inputTags,
+        team_id: this.props.team.id
+      })
+      .then(({ data }) => {
+        this.setState({ inviteResults: data });
```
Comment (ZmagoD) [inline app/javascript/src/components/InviteUsersModal/index.jsx]: you can make one sentence here:
this.setState({ inviteResults: data,  showInviteUsersResults: true});

File: app/services/client_api/invitations_service.rb [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    include UsersGenerator
+
+    def initialize(args)
+      @user = args[:user]
+      @emails = args[:emails].map(&:downcase)
+      @team = args[:team]
+      @role = args[:role]
+
+      raise ClientApi::CustomInvitationsError unless @emails && @team && @role
+      @emails && @emails.empty? { raise ClientApi::CustomInvitationsError }
```
Comment (ZmagoD) [inline app/services/client_api/invitations_service.rb]: why double checking?

File: app/services/client_api/invitations_service.rb [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+          invite_results << result
+          break
+        end
+
+        result[:email] = email
+
+        # Check if user already exists
+        user = User.find_by_email(email) if User.exists?(email: email)
+
+        result = if user.blank?
```
Comment (ZmagoD) [inline app/services/client_api/invitations_service.rb]: sorry I hate this :smile:

File: app/services/client_api/invitations_service.rb [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+                   # User does not exist
+                   handle_new_user(result, email, user)
+                 else
+                   # User exists
+                   handle_existing_user(result, user)
+                 end
```
Comment (ZmagoD) [inline app/services/client_api/invitations_service.rb]: result = handle_new_user(result, email, user) if user.blank?
result ||= handle_existing_user(result, user)

File: app/controllers/client_api/users/invitations_controller.rb [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+                   locals: { invite_results: invite_results, team: @team }
+          end
+        end
+      end
+
+      private
+
+      def check_invite_users_permission
+        @team = Team.find_by_id(params[:team_id])
+        render_403 if @team && !is_admin_of_team(@team)
```
Comment (ZmagoD) [inline app/controllers/client_api/users/invitations_controller.rb]: if we render_403 you will get a html response. Probably is better to handle this somehow different like send 422 and a message that the user doesn't have permissions

File: app/javascript/src/components/InviteUsersModal/index.jsx (hunk 9 above)
Comment (mlorb) [reply app/javascript/src/components/InviteUsersModal/index.jsx]: :+1:

File: app/services/client_api/invitations_service.rb (hunk 10 above)
Comment (mlorb) [reply app/services/client_api/invitations_service.rb]: @emails could exist, but the array could be empty.

File: app/services/client_api/invitations_service.rb (hunk 12 above)
Comment (mlorb) [reply app/services/client_api/invitations_service.rb]: če češ :+1:

File: app/controllers/client_api/users/invitations_controller.rb (hunk 13 above)
Comment (mlorb) [reply app/controllers/client_api/users/invitations_controller.rb]: :+1:

File: app/services/client_api/invitations_service.rb (hunk 10 above)
Comment (ZmagoD) [reply app/services/client_api/invitations_service.rb]: I meant on line 12 and line 13

**Pull request merged**